// bloom-filters/dist/api.js
/* file : api.ts
MIT License

Copyright (c) 2017-2020 Thomas Minier & Arnaud Grall

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var bloom_filter_1 = require("./bloom/bloom-filter");
exports.BloomFilter = bloom_filter_1.default;
var counting_bloom_filter_1 = require("./bloom/counting-bloom-filter");
exports.CountingBloomFilter = counting_bloom_filter_1.default;
var partitioned_bloom_filter_1 = require("./bloom/partitioned-bloom-filter");
exports.PartitionedBloomFilter = partitioned_bloom_filter_1.default;
var count_min_sketch_1 = require("./sketch/count-min-sketch");
exports.CountMinSketch = count_min_sketch_1.default;
var hyperloglog_1 = require("./sketch/hyperloglog");
exports.HyperLogLog = hyperloglog_1.default;
var topk_1 = require("./sketch/topk");
exports.TopK = topk_1.default;
var min_hash_1 = require("./sketch/min-hash");
exports.MinHash = min_hash_1.MinHash;
var min_hash_factory_1 = require("./sketch/min-hash-factory");
exports.MinHashFactory = min_hash_factory_1.default;
var cuckoo_filter_1 = require("./cuckoo/cuckoo-filter");
exports.CuckooFilter = cuckoo_filter_1.default;
var invertible_bloom_lookup_tables_1 = require("./iblt/invertible-bloom-lookup-tables");
exports.InvertibleBloomFilter = invertible_bloom_lookup_tables_1.default;
var cell_1 = require("./iblt/cell");
exports.Cell = cell_1.default;
